pkg/storaged/btrfs/volume.jsx

```diff
@@ -0,0 +1,192 @@
+import { DescriptionList } from "@patternfly/react-core/dist/esm/components/DescriptionList/index.js";
+import { VolumeIcon } from "../icons/gnome-icons.jsx";
+
+import { new_card, new_page, get_crossrefs, ChildrenTable, PageTable, StorageCard, StorageDescription } from "../pages.jsx";
+import { StorageUsageBar, StorageLink } from "../storage-controls.jsx";
+import { fmt_size_long, validate_fsys_label, decode_filename } from "../utils.js";
+import { btrfs_usage, btrfs_is_volume_mounted } from "./utils.jsx";
+import { dialog_open, TextInput } from "../dialog.jsx";
+import { make_btrfs_subvolume_page } from "./subvolume.jsx";
+
+const _ = cockpit.gettext;
+
+/*
+ * Udisks is a disk/block library so it manages that, btrfs turns this a bit
+ * around and has one "volume" which can have multiple blocks by a unique uuid.
+ *
+ * Cockpit shows Btrfs as following:
+ *
+ * -> btrfs subvolume
+ *    -> btrfs volume
+ *        -> btrfs device
+ *            -> block device
+ */
+export function make_btrfs_volume_page(parent, uuid) {
+    const block_devices = client.uuids_btrfs_blocks[uuid];
+    const block_btrfs = client.blocks_fsys_btrfs[block_devices[0].path];
+    const volume = client.uuids_btrfs_volume[uuid];
+    const use = btrfs_usage(client, volume);
+
+    const name = block_btrfs.data.label || uuid;
+    const btrfs_volume_card = new_card({
+        title: _("btrfs volume"),
+        next: null,
+        page_location: ["btrfs-volume", uuid],
+        page_name: name,
+        page_icon: VolumeIcon,
+        page_size: use[1],
+        component: BtrfsVolumeCard,
+        props: { block_devices, uuid, use },
+    });
+
+    const subvolumes_card = new_card({
+        title: _("btrfs subvolumes"),
+        next: btrfs_volume_card,
+        component: BtrfsSubVolumesCard,
+        props: { volume },
+    });
+
+    const subvolumes_page = new_page(parent, subvolumes_card);
+    make_btrfs_subvolume_pages(subvolumes_page, volume);
+}
+
+const BtrfsVolumeCard = ({ card, block_devices, uuid, use }) => {
+    const block_btrfs = client.blocks_fsys_btrfs[block_devices[0].path];
+    const label = block_btrfs.data.label || "-";
+
+    // Changing the label is only supported when the device is not mounted
+    // otherwise we will get btrfs filesystem error ERROR: device /dev/vda5 is
+    // mounted, use mount point. This is a libblockdev/udisks limitation as it
+    // only passes the device and not the mountpoint when the device is mounted.
+    // https://github.com/storaged-project/libblockdev/issues/966
+    const is_mounted = btrfs_is_volume_mounted(client, block_devices);
+
+    function rename_dialog() {
+        dialog_open({
+            Title: _("Change label"),
+            Fields: [
+                TextInput("name", _("Name"),
+                          {
+                              validate: name => validate_fsys_label(name, "btrfs"),
+                              value: label
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231214-115043-fd396f8d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)